platform/nuclei/ux600/platform.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright (c) Nuclei Corporation or its affiliates.
+ *
+ * Authors:
+ *   lujun <lujun@nucleisys.com>
+ */
+
+#include <libfdt.h>
+#include <sbi/riscv_asm.h>
+#include <sbi/riscv_io.h>
+#include <sbi/riscv_encoding.h>
+#include <sbi/sbi_console.h>
+#include <sbi/sbi_const.h>
+#include <sbi/sbi_platform.h>
+#include <sbi_utils/fdt/fdt_fixup.h>
+#include <sbi_utils/irqchip/plic.h>
+#include <sbi_utils/serial/sifive-uart.h>
+#include <sbi_utils/sys/clint.h>
+
+/* clang-format off */
+
+#define UX600_HART_COUNT		1
+
+#define UX600_SYS_CLK			1000000000
+
+/* Nuclei timer base address */
+#define UX600_NUCLEI_TIMER_ADDR		0x2000000
+/* The clint compatiable timer offset is 0x1000 against nuclei timer */
+#define UX600_CLINT_TIMER_ADDR		(UX600_NUCLEI_TIMER_ADDR + 0x1000)
+
+#define UX600_PLIC_ADDR			0x8000000
+#define UX600_PLIC_NUM_SOURCES		0x35
+#define UX600_PLIC_NUM_PRIORITIES	7
+
+#define UX600_UART0_ADDR		0x10013000
+#define UX600_UART1_ADDR		0x10023000
+
+#define UX600_DEBUG_UART		UX600_UART0_ADDR
+#define UX600_UART_BAUDRATE		115200
+
+/* Full tlb flush always */
+#define UX600_TLB_RANGE_FLUSH_LIMIT	0
+
+/* clang-format on */
+
+static void ux600_modify_dt(void *fdt)
+{
+	fdt_cpu_fixup(fdt);
+
+	fdt_fixups(fdt);
+}
+
+static int ux600_final_init(bool cold_boot)
+{
+	void *fdt;
+
+	if (!cold_boot)
+		return 0;
+
+	fdt = sbi_scratch_thishart_arg1_ptr();
+	ux600_modify_dt(fdt);
+
+	return 0;
+}
+
+static u32 ux600_pmp_region_count(u32 hartid)
+{
+	return 1;
+}
+
+static int ux600_pmp_region_info(u32 hartid, u32 index, ulong *prot,
+				 ulong *addr, ulong *log2size)
+{
+	int ret = 0;
+
+	switch (index) {
+	case 0:
+		*prot	  = PMP_R | PMP_W | PMP_X;
+		*addr	  = 0;
+		*log2size = __riscv_xlen;
+		break;
+	default:
+		ret = -1;
+		break;
+	};
+
+	return ret;
+}
+
+static int ux600_console_init(void)
+{
+	return sifive_uart_init(UX600_DEBUG_UART, UX600_SYS_CLK,
+				UX600_UART_BAUDRATE);
+}
+
+static int ux600_irqchip_init(bool cold_boot)
+{
+	int rc;
+	u32 hartid = current_hartid();
+
+	if (cold_boot) {
+		rc = plic_cold_irqchip_init(UX600_PLIC_ADDR,
+					    UX600_PLIC_NUM_SOURCES,
+					    UX600_HART_COUNT);
+		if (rc)
+			return rc;
+	}
+
+	return plic_warm_irqchip_init(hartid, (hartid) ? (2 * hartid - 1) : 0,
+				      (hartid) ? (2 * hartid) : -1);
+}
+
+static int ux600_ipi_init(bool cold_boot)
+{
+	int rc;
+
+	if (cold_boot) {
+		rc = clint_cold_ipi_init(UX600_CLINT_TIMER_ADDR, UX600_HART_COUNT);
+		if (rc)
+			return rc;
+	}
+
+	return clint_warm_ipi_init();
+}
+
+static u64 ux600_get_tlbr_flush_limit(void)
+{
+	return UX600_TLB_RANGE_FLUSH_LIMIT;
+}
+
+static int ux600_timer_init(bool cold_boot)
+{
+	int rc;
+
+	if (cold_boot) {
+		rc = clint_cold_timer_init(UX600_CLINT_TIMER_ADDR,
```

Comment:
Is the mmio address offset of timeval & timecmp changed in hardware ? It seems different from the previous PR.

In previous PR:
time_cmp =  UX600_RV_TIMER_ADDR + 0x08
time_val = UX600_RV_TIMER_ADDR + 0x00

Here it is complete different.

Reply:
Hi, @atishp04, Yes, our hardware changed to be able to compatiable with clint timer(start from offset 0x1000). So in software, we just adjust it. 

**Described as below:**

Addr Offset | Width | RWA | Register Name | Comment
-- | -- | -- | -- | --
0x0 | 4B | RW | mtime_lo | shadow copy of mtime | Nuclei   Timer
0x4 | 4B | RW | mtime_hi
0x8 | 4B | RW | mtimecmp_lo | shadow copy of mtimecmp for hart 0
0xC | 4B | RW | mtimecmp_hi
0xFF0 | 4B | RW | msftrst | mtimer   software reset
0xFF8 | 4B | RW | mtimectl | mtime   control
0xFFC | 4B | RW | msip | shadow   copy of msip for hart 0
0x1000 | 4B | RW | msip for hart 0 | MSIP Registers | clint   timer compatiable, offset 0x1000 from Nuclei Timer
0x1004 | 4B | RW | msip for hart 1
0x1008 | 4B | RW | msip   for hart 2
0x100C | 4B | RW | msip   for hart 3
0x1010 | 4B | RW | msip   for hart 4
… |   |   | Reserved
0x5000 | 8B | RW | mtimecmp for hart 0 | MTIMECMP Registers
0x5008 | 8B | RW | mtimecmp for hart 1
0x5010 | 8B | RW | mtimecmp   for hart 2
0x5018 | 8B | RW | mtimecmp   for hart 3
0x5020 | 8B | RW | mtimecmp   for hart 4
… |   |   | Reserved
0xCFF8 | 8B | RW | mtime | MTIME Register
… |   |   | Reserved |  

Reply:
I'm designer from Nuclei System, msftrst is actually a system control register, it uses timer's address but doesn't belong to timer. The user can trigger a system reset by writing a specific magic number to it.